### PR TITLE
Use Micronaut HTTP client for remote JWK Set validation

### PR DESCRIPTION
**Pull request by jeremyg484**
This needs some further general cleanup (haven't completed all of the JavaDoc updates, for example) and I want to add some additional tests, but I wanted to go ahead and get some early feedback before I do that additional work.

The implementation works for replacing the internal `JWKSet` URL fetching that uses JDK Http APIs with use of the Micronaut HTTP client, so that the client settings can be fully configured (to solve problems such as #1048).

My initial hesitation is around the changes that had to be made in order to propagate a JWT service name that can be matched up with the HTTP client service name, in the same manner as is done in `DefaultTokenEndpointClient`. I'm open to suggestion as to ways to improve this. Most of the impact is to internal classes and protected methods, so perhaps this is all ok.      

I did retain the existing `JwkSetFetcher#fetch` method as a potential fallback, though at the moment it will not be used and would probably need to be marked as deprecated.

### CODE REVIEW

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwkSetFetcher.java [hunk 1]
```diff
@@ -38,6 +39,15 @@ public interface JwkSetFetcher<T> {
     @Blocking
     Optional<T> fetch(@Nullable String url);
```
Comment (sdelamo) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwkSetFetcher.java:40]: should this be deprecated?

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java [hunk 2]
```diff
@@ -147,12 +148,13 @@ public boolean verify(SignedJWT jwt) throws JOSEException {
 
     /**
      * Instantiates a JWKSet for a given url.
+     * @param providerName The name of the JWKS configuration.
      * @param url JSON Web Key Set Url.
      * @return a JWKSet or null if there was an error.
      */
     @Nullable
-    protected JWKSet loadJwkSet(String url) {
```
Comment (sdelamo) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java:154]: this is a breaking change

File: security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureConfiguration.java [hunk 3]
```diff
@@ -27,6 +27,9 @@
  */
 public interface JwksSignatureConfiguration {
 
+    @NonNull
+    String getName();
```
Comment (sdelamo) [inline security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureConfiguration.java:31]: Extend from the `Named` interface instead.

File: security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/token/response/validation/DefaultOpenIdTokenResponseValidator.java [hunk 4]
```diff
@@ -147,33 +147,36 @@ protected Optional<JWT> validateClaims(@NonNull OauthClientConfiguration clientC
...
                 .validate(openIdTokenResponse.getIdToken(), null);
     }
 
     /**
-     *
+     * @param clientConfiguration The OAuth 2.0 client configuration
      * @param openIdProviderMetadata The OpenID provider metadata
      * @return A {@link JwksSignature} for the OpenID provider JWKS uri.
      */
-    protected JwksSignature jwksSignatureForOpenIdProviderMetadata(@NonNull OpenIdProviderMetadata openIdProviderMetadata) {
```
Comment (sdelamo) [inline security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/token/response/validation/DefaultOpenIdTokenResponseValidator.java:170]: this is a breaking change.

**Review by sdelamo — CHANGES_REQUESTED** [review]
It is going in the right direction. 

However,  We cannot have breaking changes. Moreover, I think it would to have a `JwksClient` API where we can encapsulate the whole `HttpClient` instantiation etc.  

We could do a `@EachBean(JwksSignatureConfiguration.class`) and create a `JwksClient` bean. 

It needs test.

**Comment by jeremyg484** [comment]
With the latest changes I have:

- Completed refactoring to retain (and deprecate) the existing API methods in favor of the new ones that support propagation of the provider name. The existing methods will still function by passing `null` for the provider name in which case they use the default HTTP client.

- Made the http client a compile only dependency and provided two implementations of a new JwksClient interface - one that uses the Nimbus library's `ResourceRetriever` API that was being used here before, and one that uses the Micronaut `HttpClient`. The `HttpClient` implementation takes precedent if the `HttpClient` is on the class path. This might be overkill, but I think it's the best way to ensure that existing apps don't break.

- Added a set of tests that simulate usage with a proxy for both implementations of JwksClient and ensures that the configured `HttpClient` settings get used. The non-proxy default settings use case is already thoroughly covered with the existing tests.

Some remaining TODOs:

- Figured out why the Java 21 CI build is failing
- Check the AOT optimizations and modify as necessary
- Add some examples to the docs about configuring the `HttpClient` settings for JWK Set retrieval.

**Comment by jeremyg484** [comment]
@sdelamo Ready for review now. Note that in this final set of changes that I've moved away from deprecating methods in `DefaultOpenIdTokenResponseValidator` (see my prior comments) and instead added a non-serialized name property to `DefaultOpenIdProviderMetadata`. This approach simplified the changes that had to be made to the AOT optimizations.

**Pull request merged**